### PR TITLE
Fix locale-dependent float parsing in C++ parser

### PR DESCRIPTION
**Pull request by yuslepukhin**
This pull request makes the ONNX parser robust to system locale settings by ensuring that float and double parsing is always locale-independent. This prevents errors when running under locales that use a comma as the decimal separator (e.g., German or French), addressing a known issue where float attributes could be misparsed. The change is implemented in a cross-platform way and is covered by new regression tests in both C++ and Python.

**Locale-independent parsing improvements:**

* Introduced `LocaleIndependentStof` and `LocaleIndependentStod` helpers in `parser.h` to parse floats/doubles in a locale-independent way, using `std::from_chars` when available or falling back to explicit "C" locale functions for portability.
* Replaced all uses of `std::stof` and `std::stod` in the parser implementation (`parser.cc` and `parser.h`) with these new helpers to ensure correct parsing regardless of the system locale. [[1]](diffhunk://#diff-5e9eb2e93be1eca35701319f90cce93b4bd64e6968a2441cc5416632fdf51263L74-R74) [[2]](diffhunk://#diff-5e9eb2e93be1eca35701319f90cce93b4bd64e6968a2441cc5416632fdf51263L587-R587) [[3]](diffhunk://#diff-5e9eb2e93be1eca35701319f90cce93b4bd64e6968a2441cc5416632fdf51263L605-R609) [[4]](diffhunk://#diff-c2cee98ca3f270d79c2ef075c84960941a092b896dbfae881092a0c79b1b0d7eL303-R423) [[5]](diffhunk://#diff-c2cee98ca3f270d79c2ef075c84960941a092b896dbfae881092a0c79b1b0d7eL317-R437)

**Testing and regression coverage:**

* Added a C++ regression test (`LocaleIndependentFloatParsing` in `parser_test.cc`) that sets a comma-decimal locale and verifies correct parsing of float literals, as well as direct testing of the new helpers.
* Added a Python regression test (`test_locale_independent_float_parsing` in `parser_test.py`) that checks float attribute parsing under a non-US locale, skipping if such a locale is not available.

**Cross-platform support:**

* Included necessary headers and platform checks to support locale-independent parsing on both Windows and POSIX systems. [[1]](diffhunk://#diff-c2cee98ca3f270d79c2ef075c84960941a092b896dbfae881092a0c79b1b0d7eR11-R24) [[2]](diffhunk://#diff-17744a29d10bdf37177898e983c0ef109f77e3e500f78bd69590f4a0934a9d32R6-R7) [[3]](diffhunk://#diff-0760da168e211cd8abada18c2060cc5b9e7dcb77638b371299cf590b46520e41R5)

### CODE REVIEW

**Review by justinchuby — APPROVED** [review]

**Comment by titaiwangms** [comment]
## Review summary

**Verdict:** The fix is real and correct — issue #8111 is genuinely resolved — but it works by a *different mechanism than the code claims*, and the live (fallback) path has two correctness gaps. Recommend changes before merge. No Critical blockers.

### Major — the `from_chars` "primary" path is dead code on every platform
`parser.h` guards the fast path with `#ifdef __cpp_lib_to_chars_floating_point`, which is **not a real C++ feature-test macro** (the standard macro is `__cpp_lib_to_chars` = `201611L`, per ISO SD-6). No conforming stdlib defines a `_floating_point` variant. Verified empirically: GCC 13.2/libstdc++ does **not** define it despite implementing floating-point `from_chars` since GCC 11. **Consequence:** every platform (GCC, Clang, MSVC) compiles the `strtof_l`/`strtod_l` fallback; the `from_chars` branch is never instantiated or tested, and the header comment ("MSVC, GCC 11+ use std::from_chars") is inaccurate.
→ Change the guard to `#ifdef __cpp_lib_to_chars` (optionally `>= 201611L`) and correct the comment.

### Major — fallback silently accepts overflow (`errno == ERANGE` not checked)
The `strtof_l`/`strtod_l` helpers only check full-consumption (`end == s.c_str() + size`), not `errno`. Input like `"1e39"` returns `inf`, fully consumed, with `errno=ERANGE`, and is **accepted**, whereas the `from_chars` path rejects it (`result_out_of_range`). Since the fallback is the *actual* live path on all platforms today, overflow literals silently become `inf` everywhere.
→ Set `errno = 0;` before the call and reject on `errno == ERANGE`. Also check `GetCLocale().loc != nullptr` and fail closed if `newlocale`/`_create_locale` failed (otherwise UB).

### Major — public-header weight + error-contract mismatch
`parser.h` is an installed public header (`CMakeLists.txt:594`). Putting non-trivial inline helpers + a Meyers-singleton static + new `<charconv>`/`<locale.h>` and a transitive `<stdexcept>` (via `ONNX_THROW`) into it leaks weight into every downstream consumer. Separately, the helpers signal errors via `ONNX_THROW` while their call sites follow the `Common::Status`/`ParseError` idiom — and the attribute conversion sites at `parser.cc:587/605/609` are **not** wrapped in `ONNX_TRY/ONNX_CATCH`, so a malformed literal throws uncaught (or `abort()`s under `ONNX_DISABLE_EXCEPTIONS`).
→ Move the helper *implementations* into `parser.cc`, leaving only declarations in the header; this isolates the throw and the heavy includes. (Cross-toolchain ODR is not a real risk within a single binary, but the public-header bloat and uncaught-throw contract are.)

### Minor
- **Behavior change:** malformed exponents `1e`, `1e+`, `1E-` (the lexer can emit these, `parser.cc:104-111`) now **throw** where `std::stof` returned `1.0`. Arguably more correct, but combined with the unwrapped call sites it becomes an uncaught throw/abort. Either fix the lexer to require exponent digits, or wrap the conversions, and add a test documenting the chosen behavior.
- **Test isolation:** both regression tests mutate process-global locale via `setlocale`. Under `pytest-xdist`/concurrent in-process runs this can pollute siblings. Prefer `self.addCleanup(...)` in the Python test (unifies early-skip and success cleanup paths), or isolate via subprocess.
- **Missing boundary tests:** overflow floats (`1e39`), and a large int64 literal exercising the `stoll` path.

### Nits
- Redundant `#ifdef _WIN32` — both branches `#include <locale.h>` identically.
- `GetCLocale()` body is duplicated verbatim in both platform branches — hoist it after the `#endif`.
- Fallback condition `end == s.c_str() || end != s.c_str() + size()` — the first operand is redundant; `end != s.c_str() + size()` suffices.
- `ONNX_THROW("..." + s)` → pass as varargs `ONNX_THROW("...", s)` to avoid a temporary.
- `result` variable name clashes semantically with `std::from_chars_result`; `<charconv>` is included unconditionally.

### Confirmed good
- `std::stol → std::stoll` is a **genuine latent Windows bug fix**: attribute INT is int64, but `long` is 32-bit on Windows LLP64, so `std::stol` truncated/threw on large attributes. LP64 Linux hid it.
- No float double-rounding (parses to `float` directly via the `float` overload / `strtof_l`).
- Full-consumption check is stricter and safer than `std::stof`'s silent trailing-character tolerance.
- **Scope is complete:** an audit of the whole `onnx/` tree found no stray `std::stof/stod/atof/strtod` left with the same locale bug.
- No regression from the classic `from_chars`-vs-`stof` differences (leading `+`/whitespace/hex) — the lexer never emits those forms.

**Top 3 to address before merge:** (1) fix the feature-test macro guard + comment, (2) add `errno == ERANGE` + null-locale checks to the fallback, (3) move implementations into `parser.cc` and/or wrap the attribute conversion sites so failures surface as `ParseError`.

<sub>🤖 Review synthesized from a multi-model review team (readability, code, critical, deep-spec, integration reviewers).</sub>

**Comment by titaiwangms** [comment]
@yuslepukhin 1.22 has already been released. Can this wait for 1.23, or you would suggest onnx to release a patch?

**Comment by yuslepukhin** [comment]
There is a customer that is hitting it, otherwise you decide according to onnx schedule. There may be other issues.

File: onnx/defs/parser.cc [hunk 1]
```diff
@@ -602,11 +721,11 @@ Common::Status OnnxParser::ParseSingleAttributeValue(AttributeProto& attr, Attri
         return ParseError("Internal error");
       case LiteralType::INT_LITERAL:
         attr.set_type(AttributeProto_AttributeType_INT);
-        attr.set_i(std::stol(literal.value));
+        attr.set_i(std::stoll(literal.value));
         break;
```
[the remark this answers is not in the document]
Comment (yuslepukhin) [reply onnx/defs/parser.cc:725]: As intended.

File: onnx/defs/parser.h [hunk 2]
```diff
@@ -11,12 +11,17 @@
...
+#include "onnx/common/common.h"
 #include "onnx/common/status.h"
 #include "onnx/onnx_pb.h"
 #include "onnx/string_utils.h"
 
 namespace ONNX_NAMESPACE {
 
+// Locale-independent string-to-float/double conversion (defined in parser.cc).
+ONNX_API float LocaleIndependentStof(const std::string& s);
+ONNX_API double LocaleIndependentStod(const std::string& s);
```
[the remark this answers is not in the document]
Comment (yuslepukhin) [reply onnx/defs/parser.h:23]: Needed in the inline functions in the header. If someone calls them, they must be exported.

**Review by justinchuby — APPROVED** [review]

File: onnx/defs/parser.cc [hunk 3]
```diff
@@ -71,7 +190,7 @@ Common::Status ParserBase::Parse(Literal& result) {
         ++next_;
       }
       ONNX_TRY {
-        static_cast<void>(std::stof(std::string(from, next_ - from)));
+        static_cast<void>(LocaleIndependentStof(std::string(from, next_ - from)));
```
Comment (justinchuby) [inline onnx/defs/parser.cc:193]: Heads-up on a behavior change: with the new full-consumption check, a malformed literal like `1e` or `1E+` (no exponent digits) now throws at attribute-set time, whereas `std::stof("1e")` previously returned `1`. This is arguably *more* correct, but if you want the error reported earlier with better context, consider requiring ≥1 digit after `e`/`E` here in the lexer so it's rejected at lex time. Non-blocking.

**Review by justinchuby — COMMENTED** [review]
Took a careful pass — this looks good and is well-scoped. 🟢 The `from_chars` primary path + `strtof_l`/`strtod_l` "C"-locale fallback is correct, all 5 `stof`/`stod` call sites are converted, full-string consumption is enforced, and the bundled `std::stol`→`std::stoll` fix is a genuine correctness improvement (32-bit `stol` on Windows LLP64 would overflow large int64 attrs). I empirically confirmed `from_chars` handles `1.`, `.5`, `1e10`, case-insensitive `inf`/`infinity`/`nan`/`-inf`, and correctly rejects `1.2.3`. The printer writes through `std::ostream` (classic locale, unaffected by `setlocale`), so serialization was already locale-safe and no round-trip gap is introduced. I also scanned the rest of the C++ sources — there are no other `stof`/`stod`/`strtod`/`atof` parse sites or narrow `std::stol` uses, so this closes the whole locale-parse class.

Two non-blocking notes:

• Exponent handling: with the new full-consumption check, a malformed literal like `1e` or `1E+` (no exponent digits) now throws at attribute-set time, whereas `std::stof("1e")` previously returned `1`. Arguably *more* correct, but you could require ≥1 digit after `e`/`E` in the lexer to reject it earlier with better context. (Also left as an inline comment on parser.cc for precise anchoring.)

• Minor: the description says the C++ tests directly test the new helpers, but `LocaleIndependentFloatParsing` only exercises them via `OnnxParser::Parse`. Consider adding a direct `EXPECT_FLOAT_EQ(LocaleIndependentStof("0.123"), 0.123f)` under the comma locale, or adjusting the wording.
